### PR TITLE
Support CSI start AlluxioFuse process in separate pod

### PR DESCRIPTION
**Pull request by ssz1997**
### What changes are proposed in this pull request?
Making CSI launch a separate pod running AlluxioFuse process, instead of launcing AlluxioFuse process in the CSI nodeserver container

### Why are the changes needed?
If nodeserver container or node-plugin pod for any reason is down, we lose Alluxio Fuse process and it's very cumbersome to bring it back. With a separate Fuse pod, CSI pod won't affect Fuse process.

Solves #14917 

### Does this PR introduce any user facing changes?
1. Removed `javaOptions` from csi section in `values.yaml`. Alluxio properties in helm chart should be organized in one place, not in `properties` and in `csi`.
2. Add property `mountInPod` in csi section. If set to `true`, Fuse process is launched in the separate pod.

### CODE REVIEW

File: integration/kubernetes/helm-chart/alluxio/values.yaml [hunk 1]
```diff
@@ -649,19 +649,21 @@ csi:
...
         limits:
-          cpu: "4"
-          memory: "8G"
+          cpu: 4
+          memory: 8G
         requests:
-          cpu: "1"
-          memory: "1G"
+          cpu: 10m
+          memory: 300Mi
```
Comment (jiacheliu3) [inline integration/kubernetes/helm-chart/alluxio/values.yaml:666]: I guess this is a FUSE daemon which is a JVM? Will a low request and high limit cause the JVM heap resize? Would it be better to allocate more resources at first? Did you get this 300Mi from some test or estimation?

**Review by jiacheliu3 — COMMENTED** [review]
Honestly I know nothing about CSI as of now, so I could only review the helm chart part :(

File: integration/kubernetes/helm-chart/alluxio/values.yaml (hunk 1 above)
Comment (ZhuTopher) [reply integration/kubernetes/helm-chart/alluxio/values.yaml:666]: If it's modal based on `mountInPod` I'd stick with the default being the greedy request, and then conditionally use some lower value in the template file (i.e: `if .Values.csi.mountInPod`).

**Review by ZhuTopher — CHANGES_REQUESTED** [review]
I wasn't able to provide much insight for the CSI changes, k8s library stuff looks fine to me. Some minor style nits here and there. Also remember to update the Helm `Chart.md` and `CHANGELOG.md`, thanks!

File: integration/kubernetes/helm-chart/alluxio/values.yaml (hunk 1 above)
Comment (ssz1997) [reply integration/kubernetes/helm-chart/alluxio/values.yaml:666]: Makes a lot of sense. Thanks

File: integration/docker/csi/alluxio/nodeserver.go [hunk 2]
```diff
@@ -87,45 +101,195 @@ func (ns *nodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublis
...
+func getFusePodObj() (*v1.Pod, error) {
+	csiFuseYaml, err := ioutil.ReadFile("/opt/alluxio/integration/kubernetes/csi/alluxio-csi-fuse.yaml")
+	if err != nil {
+		glog.V(4).Info("csi-fuse config yaml file not found")
+		return nil, status.Errorf(codes.NotFound, "csi-fuse config yaml file not found: %v", err.Error())
+	}
+	csiFuseObj, grpVerKind, err := scheme.Codecs.UniversalDeserializer().Decode(csiFuseYaml, nil, nil)
+	if err != nil {
+		glog.V(4).Info("Failed to decode csi-fuse config yaml file")
+		return nil, status.Errorf(codes.NotFound, "Failed to decode csi-fuse config yaml file.\n", err.Error())
```
[the remark this answers is not in the document]
Comment (ssz1997) [reply integration/docker/csi/alluxio/nodeserver.go:210]: Changing to `codes.Internal`. Indeed this is not a `NotFound`.

File: integration/kubernetes/helm-chart/alluxio/values.yaml (hunk 1 above)
Comment (ssz1997) [reply integration/kubernetes/helm-chart/alluxio/values.yaml:666]: Let's just keep the resources as they were. If there's no fuse process in the container, 1 cpu and 1G memory is probably excessive but I don't think it's gonna hurt

**Comment by ssz1997** [comment]
@jiacheliu3 @ZhuTopher PTAL. Thanks!

File: integration/kubernetes/helm-chart/alluxio/templates/csi/fuse-configmap.yaml [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    app: {{ $name }}
+    chart: {{ $chart }}
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
+data:
+  alluxio-csi-fuse-config: |
+    kind: Pod
+    apiVersion: v1
+    metadata:
+      name: {{ $fullName }}-fuse-
```
Comment (jiacheliu3) [inline integration/kubernetes/helm-chart/alluxio/templates/csi/fuse-configmap.yaml:33]: nit
```suggestion
      name: {{ $fullName }}-fuse
```

**Review by jiacheliu3 — APPROVED** [review]
the helm chart parts LGTM except one small nit, thanks!

File: integration/docker/csi/alluxio/nodeserver.go [hunk 4]
```diff
@@ -12,28 +12,51 @@
...
 func (ns *nodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublishVolumeRequest) (*csi.NodePublishVolumeResponse, error) {
+	if req.GetVolumeContext()["mountInPod"] == "false" {
+		glog.V(4).Infoln("Will start Fuse process in csi-node-plugin pod.")
+		return nodePublishVolumeMountProcess(req)
+	}
+	glog.V(4).Infoln("Will start Fuse process in alluxio-fuse pod.")
+	return nodePublishVolumeMountPod(req)
+}
+
+func nodePublishVolumeMountProcess(req *csi.NodePublishVolumeRequest) (*csi.NodePublishVolumeResponse, error) {
```
Comment (Binyang2014) [inline integration/docker/csi/alluxio/nodeserver.go:56]: Make it as a member function of `nodeServer`?

File: integration/docker/csi/alluxio/nodeserver.go [hunk 5]
```diff
@@ -87,45 +103,203 @@ func (ns *nodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublis
...
+	if req.GetVolumeContext()["mountInPod"] == "true" {
+		ns.mutex.Lock()
+		defer ns.mutex.Unlock()
+
+		glog.V(4).Infoln("Creating Alluxio-fuse pod and mounting Alluxio to global mount point.")
+		fusePod, err := getAndCompleteFusePodObj(ns.nodeId, req)
+		if err != nil {
+			return nil, err
+		}
+		if _, err := ns.client.CoreV1().Pods("default").Create(fusePod); err != nil {
```
Comment (Binyang2014) [inline integration/docker/csi/alluxio/nodeserver.go:193]: It should in the same namespace with `nodeserver`, not `default` namespace

File: integration/docker/csi/alluxio/nodeserver.go [hunk 6]
```diff
@@ -87,45 +103,203 @@ func (ns *nodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublis
...
+		glog.V(4).Info("csi-fuse only support pod. %v found.")
+		return nil, status.Errorf(codes.InvalidArgument, "csi-fuse only support Pod. %v found.\n%v", grpVerKind.Kind, err.Error())
+	}
+	csiFusePodObj := csiFuseObj.(*v1.Pod)
+
+	// Add nodeName to pod name for uniqueness
+	if nodeId == "" {
+		return nil, status.Errorf(codes.InvalidArgument, "nodeID is missing in the csi setup.\n%v", err.Error())
+	}
+	csiFusePodObj.Name = csiFusePodObj.Name + nodeId
```
Comment (Binyang2014) [inline integration/docker/csi/alluxio/nodeserver.go:223]: What if multi-pod mount the different volume in the same node? Will they use the same pod name?

File: integration/kubernetes/helm-chart/alluxio/CHANGELOG.md [hunk 7]
```diff
@@ -258,3 +258,9 @@
 0.6.40
 
 - Fix incorrect directory when mounting & formatting master journal volume
+
+0.6.41
+
+- Remove javaOptions under CSI
```
Comment (Binyang2014) [inline integration/kubernetes/helm-chart/alluxio/CHANGELOG.md:264]: Why remove `javaOptions`?

File: integration/docker/csi/main.go [hunk 8]
```diff
@@ -94,3 +102,19 @@ func startReaper() {
 		}
 	}()
 }
+
+func startKubeClient() (*kubernetes.Clientset, error) {
```
Comment (Binyang2014) [inline integration/docker/csi/main.go:106]: Change to `newKubeClient`?

File: integration/docker/csi/alluxio/nodeserver.go [hunk 9]
```diff
@@ -87,45 +103,203 @@ func (ns *nodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublis
...
+		if err != nil {
+			glog.V(3).Infoln("Alluxio is not mounted.")
+		}
+		if len(stdout) > 0 {
+			break
+		}
+		i++
+	}
+	if i == 10 {
+		glog.V(3).Infoln("alluxio-fuse is not mounted to global mount point in 30s.")
```
Comment (Binyang2014) [inline integration/docker/csi/alluxio/nodeserver.go:135]: Move this to `NodeStageVolume`?  After `NodeStageVolume` we should make sure the volume already ready in this node

**Review by Binyang2014 — COMMENTED** [review]
After this change. If two jobs use same volume, the these two jobs will share the same fuse daemon. This will break some isolation. And it Job A use I/O heavily, due to the thread number limitation, Job B will be impacted. The idea solution is job A and job B use different fuse daemon and can config the resource limitation based on their requirements.

File: integration/docker/csi/alluxio/nodeserver.go (hunk 4 above)
Comment (ssz1997) [reply integration/docker/csi/alluxio/nodeserver.go:56]: Sorry yes that would be better

File: integration/docker/csi/alluxio/nodeserver.go (hunk 6 above)
Comment (ssz1997) [reply integration/docker/csi/alluxio/nodeserver.go:223]: Good point. Thank you.

File: integration/kubernetes/helm-chart/alluxio/CHANGELOG.md (hunk 7 above)
Comment (ssz1997) [reply integration/kubernetes/helm-chart/alluxio/CHANGELOG.md:264]: I was thinking about putting all properties in a central place (the `properties` section). In your case, do different clients have different properties?

File: integration/docker/csi/alluxio/nodeserver.go (hunk 4 above)
Comment (ssz1997) [reply integration/docker/csi/alluxio/nodeserver.go:56]: Why should we do it? We are not using the `nodeServer` in it

**Comment by ssz1997** [comment]
> After this change. If two jobs use the same volume, then these two jobs will share the same fuse daemon. This will break the isolation we assumed before. And if Job A use I/O heavily, due to the thread number limitation, Job B will be impacted. Even more, if Job A cause fuse daemon crashed, Job B will also crashed. The idea solution is job A and job B use the different fuse daemons and job submitter can config the fuse daemon resource limitation based on their requirements.

@Binyang2014 Thank you so much for your review. I think these are good points.

However I'm wondering, if job A and B are different workloads (one I/O heavy and one not), does it make more sense to use different pv/pvc, so that they will launch separate Fuse pods and thus using different Fuse processes?

Do you prefer to just getting rid of the current way of launching Fuse processes in the nodeserver container, and launching them in different pods instead?

**Comment by Binyang2014** [comment]
> > After this change. If two jobs use the same volume, then these two jobs will share the same fuse daemon. This will break the isolation we assumed before. And if Job A use I/O heavily, due to the thread number limitation, Job B will be impacted. Even more, if Job A cause fuse daemon crashed, Job B will also crashed. The idea solution is job A and job B use the different fuse daemons and job submitter can config the fuse daemon resource limitation based on their requirements.
> 
> @Binyang2014 Thank you so much for your review. I think these are good points.
> 
> However I'm wondering, if job A and B are different workloads (one I/O heavy and one not), does it make more sense to use different pv/pvc, so that they will launch separate Fuse pods and thus using different Fuse processes?
> 
> Do you prefer to just getting rid of the current way of launching Fuse processes in the nodeserver container, and launching them in different pods instead?

For AI scenario, most of jobs sharing a  few well-known datasets such as ImageNet. Users may tune different model based on the dataset. So group/cluster admin may create one PV/PVC and all these jobs attached the same volume. Since we don't known the processing speed for each model. So maybe some model is processing faster than others. I agree admin can create different PV/PVC for different jobs, but this method seems not recommended by Kubernetes

For second question.
Yes, launching them in different pods will bring benefits for  resource isolation and platform upgrade.  There are two problems for current approach:
1. isolation: we can not cap the cpu/memory usage for each fuse daemon, if some less import jobs consuming data too fast, it will impact other critical jobs
2. Upgrading node-server will kill the fuse-daemon as well as the related jobs
So from my point of view, after migrate to pod method, we need to solve above problems.

File: integration/docker/csi/alluxio/nodeserver.go (hunk 4 above)
Comment (Binyang2014) [reply integration/docker/csi/alluxio/nodeserver.go:56]: You can ignore this if not make sense to you

File: integration/docker/csi/alluxio/controllerserver.go [hunk 10]
```diff
@@ -122,7 +126,6 @@ func (cs *controllerServer) DeleteVolume(ctx context.Context, req *csi.DeleteVol
 		glog.V(3).Infof("Invalid delete volume req: %v", req)
 		return nil, err
 	}
-	glog.V(4).Infof("Deleting volume %s", volumeID)
```
Comment (ssz1997) [inline integration/docker/csi/alluxio/controllerserver.go:125]: We don't do anything here. The log is misleading.

Comment (ZhuTopher) [reply integration/docker/csi/alluxio/controllerserver.go:125]: Does the returned value not trigger a deletion? If so why is the return type not a `CreateVolumeResponse`?
`return &csi.DeleteVolumeResponse{}, nil`

File: integration/docker/csi/alluxio/nodeserver.go [hunk 11]
```diff
@@ -87,45 +119,184 @@ func (ns *nodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublis
...
+			if len(stdout) > 0 {
+				break
+			}
+			i++
+		}
+		if i == 12 {
+			glog.V(3).Infoln("alluxio-fuse is not mounted to global mount point in 60s.")
+			return nil, status.Error(codes.DeadlineExceeded, "alluxio-fuse is not mounted to global mount point in 60s")
+		}
+	}
```
Comment (ZhuTopher) [inline integration/docker/csi/alluxio/nodeserver.go:205]: Can we make the timeout & retries configurable? I can imagine this being very varied.

Maybe we could define a readiness probe on the FUSE pod to indicate that it has finished mounting and wait on that through the K8s API, but for now the timeout is fine imo.

Comment (ZhuTopher) [inline integration/docker/csi/alluxio/nodeserver.go:205]: Style nit: Assume the exterior of the for-loop is the error case if you are waiting on a timeout with retries.

So at the top of the method you'd need:
```
if req.GetVolumeContext()["mountInPod"] == "false" {
    return &csi.NodeStageVolumeResponse{}, nil
}
```

And then this for-loop would look like the following:
```suggestion
		for i := 0; i < 12; i++ {
			time.Sleep(5 * time.Second)
			command := exec.Command("bash", "-c", fmt.Sprintf("mount | grep %v | grep alluxio-fuse", req.GetStagingTargetPath()))
			stdout, err := command.CombinedOutput()
			if err != nil {
				glog.V(3).Infoln("Alluxio is not mounted yet.")
			}
			if len(stdout) > 0 {
				return &csi.NodeStageVolumeResponse{}, nil
			}
		}
	}
	glog.V(3).Infoln("alluxio-fuse is not mounted to global mount point in 60s.")
	return nil, status.Error(codes.DeadlineExceeded, "alluxio-fuse is not mounted to global mount point in 60s")
```

File: integration/docker/csi/alluxio/controllerserver.go (hunk 10 above)
Comment (ssz1997) [reply integration/docker/csi/alluxio/controllerserver.go:125]: Actually the returned value will trigger the deletion of the pv. However that is not happening inside this function, so logging should also not be here. Plus we are not removing any data stored in Alluxio

File: integration/kubernetes/helm-chart/alluxio/templates/csi/fuse-configmap.yaml [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+          env:
+          {{- range $key, $value := .Values.fuse.env }}
+          - name: "{{ $key }}"
+            value: "{{ $value }}"
+          {{- end }}
+          securityContext:
+            privileged: true
+            capabilities:
+              add:
+                - SYS_ADMIN
```
Comment (ZhuTopher) [inline integration/kubernetes/helm-chart/alluxio/templates/csi/fuse-configmap.yaml:75]: I think it'd be good practice to have in-line comments explaining why these are necessary. Same goes for any other CSI files where this is the case.

**Review by ZhuTopher — APPROVED** [review]
Overall LGTM, thanks for all this work Shawn!

File: integration/docker/csi/alluxio/nodeserver.go [hunk 13]
```diff
@@ -87,45 +120,188 @@ func (ns *nodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublis
...
+	ns.mutex.Lock()
+	defer ns.mutex.Unlock()
+
+	glog.V(4).Infoln("Creating Alluxio-fuse pod and mounting Alluxio to global mount point.")
+	fusePod, err := getAndCompleteFusePodObj(ns.nodeId, req)
+	if err != nil {
+		return nil, err
+	}
+	if _, err := ns.client.CoreV1().Pods(os.Getenv("NAMESPACE")).Create(fusePod); err != nil {
```
Comment (Binyang2014) [inline integration/docker/csi/alluxio/nodeserver.go:185]: What if the pod already crated by previous request? Make it idempotent?

File: integration/docker/csi/alluxio/nodeserver.go [hunk 14]
```diff
@@ -87,45 +120,188 @@ func (ns *nodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublis
...
+		time.Sleep(time.Duration(timeout) * time.Second)
+		command := exec.Command("bash", "-c", fmt.Sprintf("mount | grep %v | grep alluxio-fuse", req.GetStagingTargetPath()))
+		stdout, err := command.CombinedOutput()
+		if err != nil {
+			glog.V(3).Infoln(fmt.Sprintf("Alluxio is not mounted in %v seconds.", i * timeout))
+		}
+		if len(stdout) > 0 {
+			return &csi.NodeStageVolumeResponse{}, nil
+		}
+	}
```
[the remark this answers is not in the document]
Comment (ssz1997) [reply integration/docker/csi/alluxio/nodeserver.go:209]: This will be in the next step.

**Comment by ssz1997** [comment]
@Binyang2014 If you think the PR is good to go, please approve it. Thanks!

File: integration/docker/csi/alluxio/nodeserver.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (Binyang2014) [reply integration/docker/csi/alluxio/nodeserver.go:213]: I mean seems we can simply write as flowing: 
```suggestion
command := exec.Command("bash", "-c", fmt.Sprintf("mount | grep %v | grep alluxio-fuse", req.GetStagingTargetPath()))
stdout, err := command.CombinedOutput()
if err != nil {
	glog.V(3).Infoln(fmt.Sprintf("Alluxio mount point is not ready"))
	return err
}
```

File: integration/docker/csi/alluxio/nodeserver.go [hunk 15]
```diff
@@ -87,45 +120,192 @@ func (ns *nodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublis
...
+	ns.mutex.Lock()
+	defer ns.mutex.Unlock()
+
+	glog.V(4).Infoln("Creating Alluxio-fuse pod and mounting Alluxio to global mount point.")
+	fusePod, err := getAndCompleteFusePodObj(ns.nodeId, req)
+	if err != nil {
+		return nil, err
+	}
+	if _, err := ns.client.CoreV1().Pods(os.Getenv("NAMESPACE")).Create(fusePod); err != nil {
+		if strings.Contains(err.Error(), "already exists") {
```
Comment (Binyang2014) [inline integration/docker/csi/alluxio/nodeserver.go:186]: Using http code 409 conflict for this?

**Review by Binyang2014 — APPROVED** [review]
Thanks @ssz1997 for this change

File: integration/docker/csi/alluxio/nodeserver.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (ssz1997) [reply integration/docker/csi/alluxio/nodeserver.go:213]: You mean if fuse is not ready we just return the error, and let CSI recall this method again? But the later calls will first find out that the pod already exists and directly returns success and won't check the mount point again.
Am I interpreting it right?

File: integration/docker/csi/alluxio/nodeserver.go (hunk 15 above)
Comment (ssz1997) [reply integration/docker/csi/alluxio/nodeserver.go:186]: http code exists in the http result object, and when `err` is not `nil`, we won't return the http result.

**Comment by HelloHorizon** [comment]
alluxio-bot, merge this please

File: integration/docker/csi/alluxio/nodeserver.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (Binyang2014) [reply integration/docker/csi/alluxio/nodeserver.go:213]: OK, so we'd better use pod readiness probe. If the pod not ready, we return error directly then let CSI recall this method again, if it already ready return succeed. We should not rely on if pod existed to pass the check. Is it make sense?

File: integration/docker/csi/alluxio/nodeserver.go (hunk 15 above)
Comment (Binyang2014) [reply integration/docker/csi/alluxio/nodeserver.go:186]: How about use this method https://github.com/kubernetes/apimachinery/blob/080c0c77fab5f76acfa0c31e9e59411ecc861973/pkg/api/errors/errors.go#L540?

**Pull request merged**